src/wc-navigation-root-url.js

```diff
@@ -0,0 +1,41 @@
+/**
+ * External dependencies
+ */
+import { addFilter } from '@wordpress/hooks';
+
+/**
+ * Internal dependencies
+ */
+import storage from './utils/storage';
+
+const LOCAL_STORAGE_PROP = 'wc_nav_root_url_referrer';
+
+const rootUrlStorage = storage( LOCAL_STORAGE_PROP, null );
+
+const getRedirectUrl = ( homeUrl ) => {
+	const protocolRegex = /.*?:\/\//i;
+	const slug = homeUrl.replace( protocolRegex, '' ).replace( '/', '::' );
+
+	return `https://wordpress.com/home/${ slug }`;
+};
+
+export default () => {
+	const { isWooPage, homeUrl } = window.wcCalypsoBridge;
+	const fromCalypso =
+		document.referrer.indexOf( 'https://wordpress.com' ) === 0;
+
+	if ( fromCalypso ) {
+		rootUrlStorage.set( 'calypso' );
+	} else if ( ! isWooPage ) {
+		rootUrlStorage.set( 'wcadmin' );
+	}
+
+	// Only continue to add filter if referred by Calypso
+	if ( ! ( rootUrlStorage.get() === 'calypso' ) ) {
```

Comment:
Maybe easier to read?

```suggestion
	if ( rootUrlStorage.get() !== 'calypso' ) {
```